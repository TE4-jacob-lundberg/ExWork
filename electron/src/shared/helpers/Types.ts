export interface ILink {
  url: string;
  label: string;
  type: string;
}

export interface IGameFormFields {
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  fileNames: string[];
}

export interface IGameFormErrors { [id: string]: string; }

export interface IGame {
  id: string;
  title: string;
  abbreviation: string | null;
  image: IGameImage;
  links: ILink[];
  fileNames: string[];
}

export interface IGameImage {
  url: string;
  bannerPos: IPosition;
}

export interface IPosition {
  x: string;
  y: string;
}
