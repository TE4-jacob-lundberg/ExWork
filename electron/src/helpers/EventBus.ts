/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
