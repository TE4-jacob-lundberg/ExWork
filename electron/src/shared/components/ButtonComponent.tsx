import React, { ReactNode, useState } from 'react';
import Styled from '@emotion/styled';
import { SerializedStyles } from '@emotion/core';
import ModalComponent from './ModalComponent';

interface Props {
  children: ReactNode;
  onClick: () => void;
  styling?: SerializedStyles;
  iconSize?: string;
  background?: string;
  disabled?: boolean;
  title?: string;
  type?: 'button' | 'reset' | 'submit' | undefined;
  confirm?: boolean;
  'data-testid'?: string;
}

const ButtonComponent: React.FC<Props> = function (props: Props) {
  const [showPrompt, setShowPrompt] = useState(false);

  function handleClick(): void {
    props.confirm ? setShowPrompt(true) : props.onClick();
  }

  const ButtonStyled = Styled.button`
    border: none;
    background: ${props.background || '#333'};
    color: #FFF;
    padding: 8px;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: default;
    border-radius: 5px;
    width: fit-content;
    font-size: 1.25rem;
    cursor: pointer;
    ${props.disabled && 'filter: brightness(50%);'}

    ${props.disabled || `&:hover {
      background: #666;
    }`}

    &:active {
      outline: none;
    }

    &:focus {
      outline: none;
    }

    & i {
      font-size: ${props.iconSize};
      user-select: none;
    }

    ${props.styling && props.styling.styles}
  `;
  

  return (
    <>
      <ButtonStyled 
        type={props.type}
        onClick={handleClick} 
        disabled={props.disabled} 
        title={props.title}
        data-testid={props['data-testid']}
      >
        {props.children}
      </ButtonStyled>
      {showPrompt && <ModalComponent
        title="Are you sure?"
        onClose={(): void => setShowPrompt(false)}
        onContinue={(): void => {props.onClick(); setShowPrompt(false); }}
        cancelAble
      >
        
      </ModalComponent>}
    </>
  );
};

export default ButtonComponent;
