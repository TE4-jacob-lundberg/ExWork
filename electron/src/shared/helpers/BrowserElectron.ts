// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
