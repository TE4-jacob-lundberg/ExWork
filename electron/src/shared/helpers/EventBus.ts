/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
