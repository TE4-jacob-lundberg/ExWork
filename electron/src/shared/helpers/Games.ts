import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
