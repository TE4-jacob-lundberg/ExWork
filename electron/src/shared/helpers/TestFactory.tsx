/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
