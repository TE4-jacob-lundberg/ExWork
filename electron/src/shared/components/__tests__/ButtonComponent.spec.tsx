import React from 'react';
import { fireEvent } from '@testing-library/react';

import { factory } from '../../helpers/TestFactory';
import ButtonComponent from '../ButtonComponent';

const mockedClick = jest.fn();

const testComponent = <ButtonComponent
  onClick={mockedClick}
>
  Click me
</ButtonComponent>;

describe('ButtonComponent.tsx', () => {
  it('renders correctly', () => {
    const { container } = factory(testComponent);
    
    expect(container).toMatchSnapshot();
  });

  it('reacts on click', () => {
    const { getByText } = factory(testComponent);

    fireEvent.click(getByText('Click me'));

    expect(mockedClick).toHaveBeenCalledTimes(1);
  });

  it('correctly shows an confirmation window', () => {
    const mockedConfClick = jest.fn();
    const { getByText } = factory(<ButtonComponent
      onClick={mockedConfClick}
      confirm
    >
      Click me
    </ButtonComponent>);

    fireEvent.click(getByText('Click me'));
    expect(mockedConfClick).not.toHaveBeenCalled();

    fireEvent.click(getByText('Continue'));
    expect(mockedConfClick).toHaveBeenCalledTimes(1);
  });
});
