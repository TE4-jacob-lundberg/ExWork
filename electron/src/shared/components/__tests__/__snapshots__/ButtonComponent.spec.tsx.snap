// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComponent.tsx renders correctly 1`] = `
<div>
  <button
    class="css-1b23wax"
  >
    Click me
  </button>
</div>
`;
